src/cmap/mod.rs

```diff
@@ -300,6 +300,14 @@ impl ConnectionPoolInner {
         }
     }
 
+    async fn dropped(&self, conn: Connection) {
```

Comment:
From looking at the ["Connection Monitoring and Pooling Specification" (CMAP)](https://github.com/mongodb/specifications/blob/master/source/connection-monitoring-and-pooling/connection-monitoring-and-pooling.rst), it states that the `check_in` method should be [in charge of closing connections that are "perished"](https://github.com/mongodb/specifications/blob/master/source/connection-monitoring-and-pooling/connection-monitoring-and-pooling.rst#checking-in-a-connection). That spec also [defines "perished"](https://github.com/mongodb/specifications/blob/master/source/connection-monitoring-and-pooling/connection-monitoring-and-pooling.rst#connection) to mean a number of conditions, among them "Driver-Side timeout".

Given this, I think we can consolidate this logic into the `check_in` method, and simply add another check to the `if else` chain there for connections that are still executing. Then we can remove this method and also simplify `Connection`'s `drop` implementation a little bit to just call `check_in` if the pool still exists.

Reply:
fine by me, I am not familiar with the pool logic. feel free to change accordingly. I really just needed this to be fixed and did whatever I needed to do 🙈

Reply:
@patrickfreed or we merge this, someone smarter than me with the necessary domain knowledge of the driver adds a repro to the testing suit and *then* makes those optimisations that you mention. the last week with hassling with this in production gave me some serious **ocd**

Reply:
I can just go ahead and make those changes now. Just wanted to give you the opportunity to do so if you wished :+1: